Xrf auto fit

ericdill opened the pull request:
Here's the work that I've been doing in an attempt to remove the magic nested dictionary of parameters from your code.  I am very strongly against using a dictionary that goes 3 levels deep as the major back-end for your xrf fitting.  I saw that it has leaked into scikit-xray and I would advocate for (eventually!) excising it entirely and using objects instead, due to the increased flexibility over straight dictionaries  

Also, how standardized is the xrf_parameter.json file?

Commit:
WIP: Intermediate commit to switch computers

ericdill committed:
ENH: PEP8, variable renames, default handling

- improved handling of defaults
- provided more descriptive parameter names
- edits to conform to PEP8
- simplified code for FileIO layout
- explicitly assigned models to views in top-level enaml file
- moved defaults to xrf_gui.py

ericdill committed:
WIP: Parameter refactor

ericdill committed:
WIP: Plotting works again, auto-fit still broken

ericdill commented:
The GUI does successfully open with this branch and you can plot the data, but the fitting is currently broken.  We need to figure out how to do fitting without that nested dictionary of parameters

licode commented on bubblegum/xrf/view/guessparam.enaml:
```diff
@@ -64,199 +66,29 @@ enamldef GuessParamView(DockItem): param_view:
+                                text = 'Default'
+                                clicked ::
+                                    ff.value = loop_item.default_value
+                    row_spacing = -1
```
I like the idea of using loop. However, I probably also need to output the value in a order. Maybe we can use an OrderedDict instead.

ericdill replied:
Would sorting in alphabetical order be ok? Or do you want to be able to specify exactly what order they appear in?

ericdill replied:
I ask because it's very easy to implement alphabetical sorting (in fact, de5d422 does just that)

licode commented:
xrf_parameter.json is pretty standard, I think we can stay with the structure. Thanks for all the improvements!

licode commented on bubblegum/xrf/model/guessparam.py:
```diff
@@ -43,15 +43,42 @@
+               "toop_tip={}".format(
+            self.bound_type, self.min, self.max, self.value, self.default_value,
+            self.free_more, self.adjust_element, self.e_calibration,
+            self.linear, self.name, self.description, self.tool_tip))
```
Using Parameter class makes the whole structure easy to understand. But I am still not fully convinced. Why do we need to put each item of the dict as an Atom class? This will make the code run much faster because the base level is connected to Atom?

ericdill replied:
Doing it this way provides further type-checking of the input values for xrf fitting.  Also, I can now provide gui-level access to all of the parameter fields with about 4 lines of code using auto-enaml.  Depending on how much time I have today I might be able to get that done.

The major advantage is your first sentence, the structure is easy to understand.  To me, that is paramount in GUI code, since it is so freaking easy to make GUI code unreadable

licode replied:
Got it. So this change will not penetrate to the fitting engine, but only on the GUI level, right? The whole back-end fitting is coupled with the original dictionary format.

ericdill committed:
MNT: Fitting now works again

- Reformatted Atom Parameter objects into the nested dict
  structure that scikit-xray expects (minimum viable change to
  make the gui code better without launching into a major refactor
  of xrf in scikit-xray)

ericdill committed:
ENH: Sorted parameter keys via alphabetical order

ericdill commented on bubblegum/xrf/xrf_parameter.json:
```diff
@@ -0,0 +1,168 @@
+   "adjust_element": "fixed",
+   "e_calibration":"fixed",
+   "linear":"fixed"}
+}
```
@licode I know that you liked the single-row-structure json file, but I think that putting each value on its own line makes examining diffs easier.  The use case I have in mind is checking when/who changed the default parameters

ericdill committed:
ENH: Implemented auto-enaml editor for params

(thanks @blink1073 !)

ericdill commented:
@licode check out the auto-enaml parameter editor (scroll to the bottom of the parameters view and click 'edit defaults')

ericdill committed:
ENH: Implemented .bubblegum config location

ericdill committed:
WIP: Almost have user defaults working

- Have to stop working because we need to go check out
  of the hotel.  Will try to get back to this soon

ericdill commented:
Haven't quite completed user defaults loading from `~/.bubblegum/xrf_parameters_user.json` but it's almost there... feel free to take over, as I'm not sure how much more work I'll be doing on this trip

licode commented:
brute force merging now...

licode commented:
already merged
